File: 127.word-ladder.py
#
# @lc app=leetcode id=127 lang=python3
#
# [127] Word Ladder
#

# @lc code=start
from typing import Iterable


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        def neighbors(word: str) -> Iterable:
            for i in range(len(word)):
                for j in range(26):
                    yield word[:i] + chr(j + ord('a')) + word[i + 1:]

        wordSet = set(wordList)
        if endWord not in wordSet:
            return 0
        
        queue = deque([(beginWord, 1)])
        wordSet.discard(beginWord)
        while queue:
            curr, length = queue.popleft()
            if curr == endWord:
                return length
            for word in neighbors(curr):
                if word in wordSet:
                    queue.append((word, length + 1))
                    wordSet.remove(word)
        
        return 0
        
# @lc code=end

File: 127.word-ladder.cpp
/*
 * @lc app=leetcode id=127 lang=cpp
 *
 * [127] Word Ladder
 */

// @lc code=start
class Solution {
 public:
  int ladderLength(string beginWord, string endWord, vector<string> &wordList) {
    unordered_set<string> words(wordList.begin(), wordList.end());
    if (words.count(endWord) == 0) return 0;

    unordered_set<string> beginSet, endSet;
    beginSet.insert(beginWord);
    endSet.insert(endWord);
    int level = 1, N = beginWord.size();
    while (!beginSet.empty() and !endSet.empty()) {
      if (beginSet.size() > endSet.size()) {
        beginSet.swap(endSet);
      }

      unordered_set<string> temp;
      for (auto w : beginSet) {
        for (int i = 0; i < N; ++i) {
          char t = w[i];
          for (int k = 'a'; k <= 'z'; ++k) {
            w[i] = k;
            if (endSet.count(w)) {
              return level + 1;
            }
            if (words.count(w)) {
              temp.insert(w);
              words.erase(w);
            }
          }
          w[i] = t;
        }
      }
      ++level;
      beginSet.swap(temp);
    }
    return 0;
  }
};
// @lc code=end
